contracts/test/veLON.t.sol

```diff
@@ -0,0 +1,265 @@
+    uint256 constant MAX_LOCK_TIME = 365 days;
+    uint256 public constant PENALTY_RATE_PRECISION = 10000;
+
+    uint256 public earlyWithdrawPenaltyRate = 3000;
+
+    //record the balnce of Lon and VeLon in VM
+    BalanceSnapshot.Snapshot stakerLon;
+    BalanceSnapshot.Snapshot lockedLon;
+
+    function setUp() public {
+        // Setup
+        veLon = new veLON(address(lon));
+
+        // Deal 100 ETH to user
+        deal(user, 100 ether);
+        // Mint LON to user
+        lon.mint(user, 100 * 1e18);
+        // User approve LONStaking
+        vm.prank(user);
+        lon.approve(address(veLon), type(uint256).max);
+
+        // Deal 100 ETH to otherUSer
+        deal(other, 100 ether);
+        // Mint LON to otherUser
+        lon.mint(other, 100 * 1e18);
+        // User approve LONStaking
+        vm.prank(other);
+        lon.approve(address(veLon), type(uint256).max);
+
+        // Label addresses for easier debugging
+        vm.label(user, "User");
+        vm.label(other, "Other");
+        vm.label(address(this), "TestingContract");
+        vm.label(address(lon), "LONContract");
+        vm.label(address(veLon), "veLONContract");
+    }
+
+    /*********************************
+     *          Test: setup          *
+     *********************************/
+
+    function testSetupveLON() public {
+        assertEq(veLon.owner(), address(this));
+        assertEq(address(veLon.token()), address(lon));
+        assertEq(veLon.tokenSupply(), 0);
+        assertEq(veLon.maxLockDuration(), MAX_LOCK_TIME);
+        assertEq(veLon.earlyWithdrawPenaltyRate(), 3000);
+    }
+
+    /*********************************
+     *         Test: stake           *
+     *********************************/
+    //compute the power added when staking amount added
+    function _vePowerAdd(uint256 stakeAmount, uint256 lockDuration) internal returns (uint256) {
+        uint256 unlockTime = ((block.timestamp + lockDuration) / 1 weeks) * 1 weeks; // Locktime is rounded down to weeks
+        uint256 power = (stakeAmount / MAX_LOCK_TIME) * (unlockTime - block.timestamp);
+        return power;
+    }
+
+    function _stakeAndValidate(
+        address staker,
+        uint256 stakeAmount,
+        uint256 lockDuration
+    ) internal returns (uint256) {
+        stakerLon = BalanceSnapshot.take(staker, address(lon));
+        lockedLon = BalanceSnapshot.take(address(veLon), address(lon));
+        // uint256 numberOfNftBefore = veLon.totalNFTSupply();
+        // uint256 totalVePowerBefore = veLon.totalSupply();
+        vm.startPrank(staker);
+        if (lon.allowance(staker, address(veLon)) == 0) {
+            lon.approve(address(veLon), type(uint256).max);
+        }
+        uint256 tokenId = veLon.createLock(stakeAmount, lockDuration);
+        stakerLon.assertChange(-int256(stakeAmount));
+        lockedLon.assertChange(int256(stakeAmount));
+        // assertEq(veLon.totalNFTSupply(), numberOfNftBefore + 1);
```

Comment:
I think we need to validate the NFT supply and NFT's voting power.

Reply:
The voting power for single NFT is not visible in current veLON.sol, so I left the change in the future